drivers/base/regmap/regmap-sdw-mbq.c

```diff
@@ -9,35 +10,95 @@
 #include <linux/soundwire/sdw_registers.h>
 #include "internal.h"
 
+struct regmap_mbq_context {
+	struct device *dev;
+
+	struct regmap_sdw_mbq_cfg cfg;
+
+	int val_size;
+};
+
+static int regmap_sdw_mbq_size(struct regmap_mbq_context *ctx, unsigned int reg)
+{
+	int size = ctx->val_size;
+
+	if (ctx->cfg.mbq_size) {
+		size = ctx->cfg.mbq_size(ctx->dev, reg);
+		if (size > ctx->val_size)
+			return -EINVAL;
+	}
+
+	return size;
+}
+
 static int regmap_sdw_mbq_write(void *context, unsigned int reg, unsigned int val)
 {
-	struct device *dev = context;
+	struct regmap_mbq_context *ctx = context;
+	struct device *dev = ctx->dev;
 	struct sdw_slave *slave = dev_to_sdw_dev(dev);
+	int mbq_size = regmap_sdw_mbq_size(ctx, reg);
 	int ret;
 
-	ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg), (val >> 8) & 0xff);
-	if (ret < 0)
-		return ret;
-
-	return sdw_write_no_pm(slave, reg, val & 0xff);
+	switch (mbq_size) {
+	case 4:
+		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+				      (val >> 24) & 0xff);
+		if (ret < 0)
+			return ret;
+
+		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+				      (val >> 16) & 0xff);
+		if (ret < 0)
+			return ret;
+		fallthrough;
+	case 2:
+		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+				      (val >> 8) & 0xff);
+		if (ret < 0)
+			return ret;
+		fallthrough;
+	case 1:
+		return sdw_write_no_pm(slave, reg, val & 0xff);
```

Comment:
the other point that makes me tick is that in theory regular 8-bits SDCA registers could also be deferred. But this PR adds the deferrable property only for MBQs, so not sure if we are going in the right direction if we end-up having deferrable support in two places?

Reply:
Not sure I totally follow this one, you can access regular 8-bit registers through this regmap there is no need to differentiate them the behaviour is identical.